### DISCUSSION TITLE
Android 15 | 想 hook libbase.so 中的 StringPrintf 函数，没有触发回调。

### DISCUSSION THREAD

**Discussion by Edison0716** [bytedance/android-inline-hook#70] [General]
![image](https://github.com/bytedance/android-inline-hook/assets/8669094/7b4cedb4-441b-446e-9c39-79549865c0e4)
![image](https://github.com/bytedance/android-inline-hook/assets/8669094/653efcd4-670e-4baa-bdfb-bbe3acc9f8c8)

_ZN7android4base12StringPrintfEPKcz

**Comment by Edison0716** [comment]
![image](https://github.com/bytedance/android-inline-hook/assets/8669094/fd1fb661-90c9-49af-814f-53cb1c42f363)
hook 是成功的 但是代理函数没有执行调用。

**Comment by Edison0716** [comment]
libart.so 的函数 -> libbase.so 的函数。 此时想要 hook libbase.so 的函数 然后做一些行为。

**Reply by Colibrow** [reply]
可以先用frida hook一下目标函数确认是否走到了这个逻辑

**Comment by Edison0716** [comment]
![image](https://github.com/bytedance/android-inline-hook/assets/8669094/4bfc394d-03bf-4b48-938d-4c93b14b627d)

手动触发 确实会触发 hook proxy 的调用。

但是源码侧 确实看起来调用了

![image](https://github.com/bytedance/android-inline-hook/assets/8669094/68b58a86-06f4-4921-b292-314aa6417e7d)

但是 就是没有 Proxy Func 的回调。

**Reply by Edison0716** [reply]
@Colibrow

**Reply by Colibrow** [reply]
可以贴一下你的hook 代码和proxy 函数看一下

**Reply by Edison0716** [reply]
```

#define SYMBOL_STRING_PRINTF "_ZN7android4base12StringPrintfEPKcz"

const char *getStringPrintfFunctionName() {
    return SYMBOL_STRING_PRINTF;
}

namespace hookThreadSuspendAbortV15 {
    jobject callbackObj = nullptr;
    void *originalStringPrintf = nullptr;

    typedef void *(*StringPrintf_t)(const char *format, ...);

    bool checkFormat(const char *format);

    void *proxyStringPrintfFunc(const char *format, ...) {
        va_list args;
        va_start(args, format);
        __android_log_print(ANDROID_LOG_ERROR, LOG_TAG_THREAD_SUSPEND_HOOK_V15, "hit the hook point.");
        void *originCallback;
        if (checkFormat(format)) {
            using namespace kbArt;
            WrappedSuspend1Barrier *wrapped_barrier = va_arg(args, WrappedSuspend1Barrier*);
            if (wrapped_barrier != nullptr) {
                if (wrapped_barrier->barrier_.load() != 0) {
                    if (callbackObj != nullptr) {
                        // call the Java callback function.
                        NotifyHandleThreadSuspendTimeout::triggerSuspendTimeout(callbackObj);
                    }
                    // set the barrier to 0 to avoid the abort()...
                    wrapped_barrier->barrier_.store(0);
                    __android_log_print(ANDROID_LOG_ERROR, LOG_TAG_THREAD_SUSPEND_HOOK_V15, "set 0");
                }
            }
            originCallback = ((StringPrintf_t) originalStringPrintf)(format, args);
        } else {
            originCallback = ((StringPrintf_t) originalStringPrintf)(format, args);
        }
        va_end(args);
        return originCallback;
    }

    void fixNativeThreadSuspend(JNIEnv *env, jobject callback) {
        BaseInlineHook baseInlineHook = BaseInlineHook(env);
        baseInlineHook.callbackObj = env->NewGlobalRef(callback);
        callbackObj = baseInlineHook.callbackObj;
        baseInlineHook.setupHook(TARGET_LIB_BASE,
                                 getStringPrintfFunctionName(),
                                 (void *) proxyStringPrintfFunc,
                                 (void **) &originalStringPrintf);
    }

```

```
    bool setupHook(const char *lib_name, const char *symName, void *newFunc, void **originalFunc) {
        releaseHook();
        void *stubFunc = shadowhook_hook_sym_name(lib_name, symName, newFunc, originalFunc);
        return handleHookSetupFailure(stubFunc);
    }
```

目的是 抑制线程挂起超时崩溃的逻辑。

`libart.so` 中的 `SuspendThread` 函数内部会监控挂起是否超时，如果超时会发送 `abort` 信号。

![image](https://github.com/bytedance/android-inline-hook/assets/8669094/e6d9944c-7cf2-4884-bdd1-22281731b596)

libart.so 中的 `StringPrintf` 是 未定义的外部引用，实际在 libbase.so 中。

然后我用 Android 15 的模拟器，模拟线程挂起超时，

![image](https://github.com/bytedance/android-inline-hook/assets/8669094/8c26ac08-390e-48cd-9fa9-cb3e3afb4c08)

崩溃前 并没有调用 StringPrintf 的 proxy 函数。


https://cs.android.com/android/platform/superproject/main/+/main:art/runtime/thread_list.cc;l=1155?q=thread_list.cc
@Colibrow

**Reply by Edison0716** [reply]
改用plt hook了，亲测可以，说明源码侧的确调用了。

**Reply by Colibrow** [reply]
确定PLTHOOK之后在SuspendThread的函数里面调用了StringPrintf吗，我看你发的截图里面
> No such thread for suspend
> SuspendThreadByPeer failed for unattached thread

这两个应该会直接跳出原有的Suspend逻辑吧
还有就是对于std::string的返回值不能用void*接收，需要遵循C++的RVO返回值优化，可以参考这个文章https://developer.arm.com/documentation/den0024/a/The-ABI-for-ARM-64-bit-Architecture/Register-use-in-the-AArch64-Procedure-Call-Standard/Indirect-result-location

**Reply by Edison0716** [reply]
![image](https://github.com/user-attachments/assets/0fc9a113-e162-499c-91b8-cae374f56a1e)

Plt hook 亲测可以。

![image](https://github.com/user-attachments/assets/5659c05a-59b1-42fe-9081-e9dba2f47109)

![image](https://github.com/user-attachments/assets/f8b67022-e687-4bcc-ba03-55ce5d9a87ce)

![image](https://github.com/user-attachments/assets/f508b013-ce76-4a0b-9307-c4efb5342788)

--------------- 上述是 plt

关于 inline hook 我返回值类型也尝试返回 std::string 但是仍然无法调用代理函数。。

**Reply by Edison0716** [reply]
@Colibrow

**Reply by Colibrow** [reply]
ShadowHook目前使用xdl查找符号，针对这个case，系统中存在`/system/lib64/libbase.so`和`/apex/com.android.art/lib64/libbase.so`两个同名的so，并且都有StringPrintf符号，在Inline查符号的时候找错了so，可以指定hook的libName为`/apex/com.android.art/lib64/libbase.so`解决，并且建议在Inline和PLT都满足需求的情况下，优先使用PLTHOOK

**Reply by Edison0716** [reply]
我试试

**Reply by Edison0716** [reply]
@Colibrow 没问题了 谢谢帮助
